### PR TITLE
G5 save recipes

### PR DESCRIPTION
**Pull request by jdselby24**

### CODE REVIEW

File: src/Validator/Validator.php [hunk 1]
```diff
@@ -14,7 +14,8 @@ abstract class Validator {
  * 
  * @return string validated ingredient
  */
-    public static function validateIngredient(string $ingredient) :string {
+    public static function validateIngredient(string $ingredient) :string
```
Comment (ashleycoles) [inline src/Validator/Validator.php]: The param type no longer matches whats in the docblock

File: src/Validator/UserIngredientValidator.php [hunk 2]
```diff
@@ -15,7 +15,8 @@ abstract class UserIngredientValidator {
 	*
 	* @return array of validated user-inputted ingredients
```
Comment (ashleycoles) [inline src/Validator/UserIngredientValidator.php]: This is no longer true, the method can return an array or a string

File: src/Resource/Fridge.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
+<?php
+
+namespace GRUB\Resource;
+
+class Fridge {
```
Comment (ashleycoles) [inline src/Resource/Fridge.php]: Missing class docblock

File: src/Resource/FormDataHandler.php [hunk 4]
```diff
@@ -1,5 +1,7 @@
 <?php
```
Comment (ashleycoles) [inline src/Resource/FormDataHandler.php]: This class looks like it is only used for validating the ingredients form, would it make more sense to put it in the ingredients namespace?

File: src/Resource/FormDataHandler.php [hunk 5]
```diff
@@ -14,7 +16,8 @@ class FormDataHandler
      * @param $formData post from the form
      * @return array of ingredients from the user, validated
      */
-    public function processData(array $formData) :array{
+    public function processData(array $formData) :array
```
Comment (ashleycoles) [inline src/Resource/FormDataHandler.php]: Make sure return typehinting is consistent with the rest of the app. processData(): array instead

File: src/Recipe/RecipeEntity.php [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        $html .= "<a href='$this->link'><button>Link to Recipe</button></a>";
+        $html .= "</div>";
+        $html .= "</div>";
+        $html .= "<div class='right'>";
+        $html .= "<h5>$this->title</h5>";
+        $html .= "<p>Ingredients: $this->ingredients </p>";
+        $html .= "</div>";
+        $html .= "</div>";
+        return $html;
+    }
```
Comment (ashleycoles) [inline src/Recipe/RecipeEntity.php]: This is not very DRY!

File: src/Recipe/RecipeEntity.php [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+    /**
+     * Limits the number of ingredients to 21
+     *
+     * @param [string] $phrase The string of ingredients to be trimmed
```
Comment (ashleycoles) [inline src/Recipe/RecipeEntity.php:108]: Remove square brackets in docblock

File: src/Recipe/RecipeEntity.php [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+              $this->title = html_entity_decode($title);
+              $this->link = $link;
+
+            if(strlen($imageURL) == 0) {
+                $this->imageURL = "img/can.jpg";
+            } else {
+                $this->imageURL = $imageURL;
+            }
+
+            $this->ingredients = $this->limit_ingredients($ingredients);
```
Comment (ashleycoles) [inline src/Recipe/RecipeEntity.php]: Check indents

File: src/Recipe/RecipeDBHydrator.php [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+    }
+
+    /**
+     * Function that retrieves the recipe objects from the DB
+     *
+     * @return array Array of recipe objects
+     */
+    public function getRecipesFromDB(): array
+    {
+        $recipesOut = [];
```
Comment (ashleycoles) [inline src/Recipe/RecipeDBHydrator.php]: This is defined on line 36 but not used until line 46, define it closer to where it gets used. Not a big deal in this case, but over time it's common for variable declarations to happen miles away from where they get used (as people add code in between) so it's best to keep on top of it.

File: src/Recipe/RecipeDBHydrator.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+        $statement = "SELECT `title`, `link`, `imageURL`, `ingredients` FROM `recipes`;";
+        $query = $this->db->prepare($statement);
+        $query->setFetchMode(PDO::FETCH_ASSOC);
```
Comment (ashleycoles) [inline src/Recipe/RecipeDBHydrator.php]: Why use FETCH_ASSOC and manual object creation? You used FETCH_CLASS elsewhere so I would definitely use that here.

File: src/Recipe/RecipeDBHydrator.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        $query->execute();
+        $savedRecipes =  $query->fetchAll();
+
+        foreach($savedRecipes as $recipe) {
+            $recipeObj = new RecipeEntity($recipe['title'], $recipe['link'], $recipe['imageURL'], $recipe['ingredients']);
+            array_push($recipesOut, $recipeObj);
+        }
```
Comment (ashleycoles) [inline src/Recipe/RecipeDBHydrator.php]: This becomes redundant when you switch to using fetch_class

File: savedRecipes.php [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+$htmlOut = "";
+$recipeDBHydrator = new RecipeDBHydrator(GRUB\Resource\Db::getDB());
+$recipes = $recipeDBHydrator->getRecipesFromDB();
+if(count($recipes) != 0) {
+foreach($recipes as $recipe) {
+$htmlOut .=  $recipe->generateHTMLSaved();
+}
+} else {
+$htmlOut =  "<h3>You have no recipes saved</h3>";
+}
```
Comment (ashleycoles) [inline savedRecipes.php]: Fix indenting

File: recipes.php [hunk 13]
```diff
@@ -1,14 +1,17 @@
...
-
+session_start();
 require_once "vendor/autoload.php";
 
-use GRUB\Curl;
-use GRUB\RecipeHydrator;
-use GRUB\Validator;
+use GRUB\Resource\Curl;
+use GRUB\Recipe\RecipeHydrator;
+use GRUB\Validator\Validator;
```
Comment (ashleycoles) [inline recipes.php]: This doesn't seem to be used

File: recipes.php [hunk 14]
```diff
@@ -1,14 +1,17 @@
...
+$message = "";
 
 if($_POST != []) {
+    $_SESSION['ingredients'] = $_POST;
     $htmlOut = '';
-    $formDataHandler = new GRUB\FormDataHandler();
+    $formDataHandler = new GRUB\Resource\FormDataHandler();
     $ingredients = $formDataHandler->processData($_POST);
     $request = new Curl($ingredients);
     $recipeHydrator = new RecipeHydrator($request);
```
Comment (ashleycoles) [inline recipes.php]: Lots of code repetition here, try and think about ways to minimise it.

File: recipes.php [hunk 15]
```diff
@@ -21,7 +24,31 @@
...
+            foreach($recipes as $recipe) {
+                $htmlOut .=  $recipe->generateHTML();
+            }
+        } else {
+            $htmlOut =  "<h1>No recipes found, please select different ingredients</h1>";
+        }
+    } else {
+        header("Location: index.php?message=Please%20select%20some%20ingredients");
+    }
+}
```
Comment (ashleycoles) [inline recipes.php]: Repetition

**Review by ashleycoles — CHANGES_REQUESTED** [review]

File: recipes.php (hunk 13 above)
Comment (sudokufan) [reply recipes.php]: you're right! Removed

File: src/Resource/Fridge.php (hunk 3 above)
Comment (sudokufan) [reply src/Resource/Fridge.php]: doc block added

File: src/Recipe/RecipeDBHydrator.php (hunk 11 above)
Comment (jdselby24) [reply src/Recipe/RecipeDBHydrator.php]: Can't use fetch class due to construct needing params

File: src/Validator/Validator.php (hunk 1 above)
Comment (mgallagher56) [reply src/Validator/Validator.php]: Fixed

File: savedRecipes.php (hunk 12 above)
Comment (sudokufan) [reply savedRecipes.php]: fixed!

File: src/Resource/FormDataHandler.php (hunk 5 above)
Comment (sudokufan) [reply src/Resource/FormDataHandler.php]: done

File: src/Validator/UserIngredientValidator.php (hunk 2 above)
Comment (sudokufan) [reply src/Validator/UserIngredientValidator.php]: fixed, will this suffice?

File: src/Recipe/RecipeDBHydrator.php (hunk 11 above)
Comment (ashleycoles) [reply src/Recipe/RecipeDBHydrator.php]: You can use the third param to parse params to the constructor: https://www.php.net/manual/en/pdostatement.fetchall.php

You can also use the __set magic method in conjunction with PDO::FETCH_PROPS_LATE to handle mismatched property names https://stackoverflow.com/questions/8898794/using-pdofetch-class-with-magic-methods

File: src/Validator/UserIngredientValidator.php (hunk 2 above)
Comment (ashleycoles) [reply src/Validator/UserIngredientValidator.php]: Almost! The 'official' way of showing multiple return types is like this:

* @return array|string Array of stuff or a string of stuff

File: src/Resource/FormDataHandler.php (hunk 4 above)
Comment (sudokufan) [reply src/Resource/FormDataHandler.php]: done

File: tests/src/Resource/FormDataHandler.php [hunk 16]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+require_once('../../../src/Resource/FormDataHandler.php');
```
Comment (ashleycoles) [inline tests/src/Resource/FormDataHandler.php:3]: Test does not work

**Review by ashleycoles — APPROVED** [review]

**Pull request merged**